clientv3/grpc_dial_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/pkg/testutil"
+
+	"sync"
+
+	"google.golang.org/grpc"
+)
+
+// TestBalancerDoNotBlockOnClose ensures that balancer and grpc don't deadlock each other
+// due to rapid open/close conn. The deadlock causes balancer.Close() to block forever.
+// See issue: https://github.com/coreos/etcd/issues/7283 for more detail.
+func TestBalancerDoNotBlockOnClose(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	kcl := NewKillConnListener([]string{"127.0.0.1:8081", "127.0.0.1:8082"})
+	kcl.Start()
+	defer kcl.Close()
+
+	timer := time.After(10 * time.Second)
+	for {
+		select {
+		case <-timer:
+			return
+		default:
+		}
+		conn, err := grpc.Dial("127.0.0.1:8081", grpc.WithInsecure(),
+			grpc.WithBalancer(newSimpleBalancer([]string{"127.0.0.1:8081", "127.0.0.1:8082"})))
+		if err != nil {
+			t.Fatal(err)
+		}
+		// sleep a bit to allow connections to be established.
+		time.Sleep(100 * time.Millisecond)
+		closec := make(chan struct{})
+		// Close() might block on balancer.Close()
+		// if balancer and grpc deadlock each other.
+		go func() {
+			defer close(closec)
+			conn.Close()
+		}()
+		select {
+		case <-closec:
+		case <-time.After(time.Second):
+			t.Errorf("conn.Close() timeout")
+		}
+	}
+}
+
+// killConnListener listens incoming conn and kills it immediately.
+type killConnListener struct {
+	wg    sync.WaitGroup
+	eps   []string
+	donec chan struct{}
+
+	// mu protects lns
+	mu  sync.Mutex
+	lns []net.Listener
+}
+
+func NewKillConnListener(eps []string) *killConnListener {
+	return &killConnListener{eps: eps, donec: make(chan struct{})}
+}
+
+func (kcl *killConnListener) Start() {
+	for _, ep := range kcl.eps {
+		kcl.wg.Add(1)
+		go func(endpoint string) {
+			defer kcl.wg.Done()
+			kcl.listen(endpoint)
+		}(ep)
+	}
+}
+
+func (kcl *killConnListener) listen(url string) {
+	ln, _ := net.Listen("tcp", url)
```

Comment:
use unix listeners or an arbitrary port via `:0` or this will break on CI from port conflicts